homeassistant/components/konnected/config_flow.py

```diff
@@ -0,0 +1,732 @@
+        self.port = None
+        self.model = KONN_MODEL
+        self.device_id = None
+
+        # data above builds the config entry
+        # data below maintains state between steps
+        self.io_cfg = {}
+        self.binary_sensors = []
+        self.sensors = []
+        self.switches = []
+        self.active_cfg = None
+
+    @property
+    def cached_config(self):
+        """Retrieve cached config for device."""
+        if self.hass.data.get(DOMAIN) and self.hass.data[DOMAIN].get("config_data"):
+            return self.hass.data[DOMAIN]["config_data"].get(self.device_id)
+        return None
+
+    @callback
+    def async_cache_config(self, config):
+        """Cache the device config for shared usage."""
+        if DOMAIN not in self.hass.data:
+            self.hass.data[DOMAIN] = {}
+            self.hass.data[DOMAIN]["config_data"] = {}
+
+        self.hass.data[DOMAIN]["config_data"][self.device_id] = config
+
+    async def async_step_init(self, user_input=None):
+        """Needed in order to not require re-translation of strings."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_import(self, import_info):
+        """Import a configuration.yaml config.
+
+        This flow is triggered by `async_setup` for configured panels.
+        Triggered for any panel that does not have a config entry
+        yet (based on device_id).  If the cfg entry can't be made
+        we'll cache the data in hass.data[DOMAIN]["config_data"] for
+        other flows to reference.
+        """
+        _LOGGER.debug(import_info)
+        try:
+            device_config = DEVICE_SCHEMA_YAML(import_info)
+
+        except vol.Invalid as err:
+            _LOGGER.error(
+                "Cannot import config..%s", humanize_error(import_info, err),
+            )
+            return self.async_abort(reason="unknown")
+
+        # swap out pin for zones in a io config
+        def pins_to_zones(config):
+            for zone in config:
+                if zone.get(CONF_PIN):
+                    zone[CONF_ZONE] = PIN_TO_ZONE[zone[CONF_PIN]]
+                    del zone[CONF_PIN]
+
+        if device_config.get(CONF_BINARY_SENSORS):
+            pins_to_zones(device_config[CONF_BINARY_SENSORS])
+
+        if device_config.get(CONF_SENSORS):
+            pins_to_zones(device_config[CONF_SENSORS])
+
+        if device_config.get(CONF_SWITCHES):
+            pins_to_zones(device_config[CONF_SWITCHES])
+
+        self.device_id = self.context["device_id"] = device_config["id"]
+        try:
+            self.host = device_config[CONF_HOST]
+            self.port = device_config[CONF_PORT]
+
+        except KeyError:
+            # cache config and wait for user input or discovery to provide host info
+            self.async_cache_config(device_config)
+            return await self.async_step_user()
+
+        # create the config entry
+        return await self.async_create_or_update_entry(device_config)
```

Comment:
We still want to validate the connection even if the user provides the host and port in the config yaml.

Reply:
I struggled with the correct move here because of the device offline use case. 

If we block the import for an offline device - the device cfg entry won't be created until the the next time HA starts. 
If we import and create a config entry, the standard config setup process will retry the connection until it can be made.

It's possible they entered the wrong host info but I assume someone editing the yaml must have good reason.

Reply:
I think it's ok to require the device to be online during the first config flow import.

It's only required the first time a unique config entry is created. After that the config entry will just load without going through the whole flow since existing entries will abort the flow.

Inside `async_setup_entry` in the component we can raise `ConfigEntryNotReady` to try retry on offline device.